lib/api/api.go

```diff
@@ -497,7 +498,7 @@ func (s *service) CommitConfiguration(from, to config.Configuration) bool {
 	// No action required when this changes, so mask the fact that it changed at all.
 	from.GUI.Debugging = to.GUI.Debugging
 
-	if to.GUI == from.GUI {
+	if cmp.Equal(to.GUI, from.GUI) {
```

Comment:
What's the reason for this change?

Had a quick look at it's doc, and this seems like a dealbreaker:

> If no custom equality functions are used and no Equal method is defined, equality is determined by recursively comparing the primitive kinds on both values, much like [reflect.DeepEqual](https://pkg.go.dev/reflect#DeepEqual). Unlike [reflect.DeepEqual](https://pkg.go.dev/reflect#DeepEqual), unexported fields are not compared by default; they result in panics unless suppressed by using an [Ignore](https://pkg.go.dev/github.com/google/go-cmp/cmp#Ignore) option (see [github.com/google/go-cmp/cmp/cmpopts.IgnoreUnexported](https://pkg.go.dev/github.com/google/go-cmp@v0.6.0/cmp/cmpopts#IgnoreUnexported)) or explicitly compared using the [Exporter](https://pkg.go.dev/github.com/google/go-cmp/cmp#Exporter) option.

Probably there will newer be unexported fields, but it's not impossible, and we definitely don't want panics if it ever happens.

Also on a second look:

> It is intended to only be used in tests, [...]

Reply:
Oh, my bad. The reason is this:

```
lib/api/api.go:501:5: invalid operation: to.GUI == from.GUI (struct containing []byte cannot be compared)
```

So _something_ else than `==` is needed for this to work. Is [reflect.DeepEqual](https://pkg.go.dev/reflect#DeepEqual) the correct solution to this, then?

Reply:
(This also snuck into the main PR in https://github.com/syncthing/syncthing/pull/9175/commits/844516eca2735051a3faec04c529f35c4e29d08f, but I'll amend that too with whatever resolution we reach here)

Reply:
Perhaps this part of [reflect.DeepEqual](https://pkg.go.dev/reflect#DeepEqual) could be problematic:

>Note that a non-nil empty slice and a nil slice (for example, []byte{} and []byte(nil)) are not deeply equal.

But it seems to at least pass all the tests both in this PR and in https://github.com/syncthing/syncthing/pull/9175.

Reply:
Oh that's indeed odd - usually a nil and zero length slice behave just the same in go. However likely not an issue as we do the same thing everywhere. If you wanted to be sticklerish you could add an `Equal` method, manually compare the two slices and then set them to `nil` for `reflect.DeepEqual`. However imo that's overkill.

Reply:
Ok, I'll just use `reflect.DeepEqual` for now then. Done in caad2ff7b8ef3fd8a49055650ea0b3b1b444f666.